sympy/parsing/sympy_parser.py

```diff
@@ -507,6 +509,95 @@ def implicit_multiplication_application(result, local_dict, global_dict):
 
     return result
 
+def find_closing(string,pos,opening,closing):
+    # find the next closing
+    ic=string.index(closing,pos)
+    # print("ic",ic)
+    # find the next opening in the space between if there is one
+    try:
+        io=string.index(opening,pos+1,ic-1)
+        # print("io",io)
+        # find the closing to the next (leftmost) opening
+        ioc=find_closing(string,io,opening,closing)
+        # and start from there
+        return string.index(closing,ioc+1)
+    except ValueError:
+        #if there is no opening the next closing was the matching one
+        return ic
+
+
+def auto_indexed_base(tokens, local_dict, global_dict):
+    """Inserts calls to ``IndexedBase`` for undefined indexed variables."""
+    result = []
+    prevTok = (None, None)
+    n=len(tokens)
+    local_dict=copy.copy(local_dict)
+    tokens.append((None, None))  # to provide a last nextTok
+    tokvals=[t[1] for t in tokens]
+
+    for i in range(n):
+        tok=tokens[i]
+        tokNum, tokVal = tok
+        name = tokVal
+        nextTok=tokens[i+1]
+        nextTokNum, nextTokVal = nextTok
+        if tokNum == NAME and nextTokVal=='[':
+            # find closing ']'
+            # todo:
+            # we could catch the possible exception if no closing ']' is found
+            # to make the resulting errormessage more intelligible
+            ic=find_closing(tokvals,i+1,'[',']')
+            inner_tokens=tokens[i+2:ic]
+            # print('inner_tokens',inner_tokens)
+            # The element between '[' and ']' should be either a number or a tuple
+            # The rank of the IndexedBase will be 1 or the length of the tuple
+            # To determine the kind of the inner expression we call sympify recursively
+            innerExp=sympify(untokenize(inner_tokens))
+
+            if isinstance(innerExp,tuple):
+                rank=len(innerExp)
+            else:
+                rank=1
+            # define the (dummy)indices and the shape.
+            # todo:
+            # we should make sure that the index variables are not defined yet)
+            my_shape=symbols(["dummy_"+str(i) for i in range(rank)],cls=Idx)
+            new=IndexedBase(name,shape=my_shape)
```

Comment:
You should just leave the shape as None (the default).

Reply:
Maybe I'm missing why you are adding a shape, but this should simplify this code a lot. For instance, there is no need to do the stuff with the locals dictionary.